File: mergesort.java
// java program for merge sort
import java.util.*;
class MergeSort {

	//merge sort function
	public static void mergeSort(int[] a, int n) {
		if (n < 2) {
			return;
		}
		int mid = n / 2; 
		int[] l = new int[mid]; 
		int[] r = new int[n - mid]; 

		for (int i = 0; i < mid; i++) {
			l[i] = a[i];
		}
		for (int i = mid; i < n; i++) {
			r[i - mid] = a[i];
		}
		mergeSort(l, mid);
		mergeSort(r, n - mid);

		merge(a, l, r, mid, n - mid);
	}
	public static void merge(
			int[] a, int[] l, int[] r, int left, int right) {

		int i = 0, j = 0, k = 0;
		
		while (i < left && j < right) {
			if (l[i] <= r[j]) {
				a[k++] = l[i++];
			}
			else {
				a[k++] = r[j++];
			}
		}
		while (i < left) {
			a[k++] = l[i++];
			a[k++] = l[i--]; 
		}
		while (j < right) {
			a[k++] = r[j++];
		}
	}

	
	static void printArray(int arr[])
	{
		
		for (int i = 0; i <  arr.length; ++i)
			System.out.print(arr[i] + " ");
		System.out.println();
	}

	// Driver code
	public static void main(String args[])
	{
		// taking input
		Scanner sc=new Scanner(System.in);
		
		int p=sc.nextInt();
		
		int[] arr = new int[p];  
		for(int i=0; i<p; i++)  
        {  
        	arr[i]=sc.nextInt();  
        }

		System.out.println("Given Array: ");
		printArray(arr);
		mergeSort(arr, p);
		System.out.println("\nSorted array");
		printArray(arr);
	}
}

/* This code is contributed by shakti Singh */

File: Merge_Sort.cpp
#include <bits/stdc++.h>

using namespace std;

int *temp;

// merge 
void merge(int arr[], int left, int right, int mid)
{
    int k = left, i, j;

    for(i = left, j = mid + 1; i <= mid && j <= right; k++)
    {
        if(arr[i] <= arr[j])
        {
            temp[k] = arr[i];
            i++;
        }
        else
        {
            temp[k] = arr[j];
            j++;
        }
    }

    while(i <= mid)
    {
        temp[k] = arr[i];
        i++;
        k++;
    }

    while(j <= right)
    {
        temp[k] = arr[j];
        j++;
        k++;
    }

    for(i = left; i <= right; i++)
        arr[i] = temp[i];
}


// Divide array into halves
void divide(int arr[], int s, int e)
{
    if(s < e)
    {
        int mid = s + (e - s) / 2;

        divide(arr, s, mid);
        divide(arr, mid + 1, e);

        merge(arr, s, e, mid);
    }
}

void Merge_Sort(int arr[], int size)
{
    temp = new int[size];
    divide(arr, 0, size - 1);
    delete [] temp;
}

// Function to print elements of array
void printArray(int arr[], int size)
{
    for(int i = 0; i < size; i++)
        cout << arr[i] << " ";

    cout << endl;
}

int main()
{
    int arr[] = {2, 4, 3, 1, 6, 8, 4};
    Merge_Sort(arr, 7);
    printArray(arr, 7);
    return 0;
}

// Output
// 1 2 3 4 4 6 8
